docs/protocols/java_object_graph_spec.md

```diff
@@ -0,0 +1,541 @@
+# Fury Java Serialization Specification
+
+## Spec overview
+
+Fury Java Serialization is an automatic object serialization framework that supports reference and polymorphism. Fury
+will
+convert an object from/to fury java serialization binary format. Fury has two core concepts for java serialization:
+
+- **Fury Java Binary format**
+- **Framework to convert object to/from Fury Java Binary format**
+
+The serialization format is a dynamic binary format. The dynamics and reference/polymorphism support make Fury flexible,
+much more easy to use, but
+also introduce more complexities compared to static serialization frameworks. So the format will be more complex.
+
+Here is the overall format:
+
+```
+| fury header | object ref meta | object class meta | object value data |
+```
+
+The data are serialized using little endian byte order overall. If bytes swap is costly for some object,
+Fury will write the byte order for that object into the data instead of converting it to little endian.
+
+## Fury header
+
+Fury header consists starts one byte:
+
+```
+|     4 bits    | 1 bit | 1 bit | 1 bit  | 1 bit |          optional 4 bytes          |
++---------------+-------+-------+--------+-------+------------------------------------+
+| reserved bits |  oob  | xlang | endian | null  | unsigned int for meta start offset |
+```
+
+- null flag: 1 when object is null, 0 otherwise. If an object is null, other bits won't be set.
+- endian flag: 1 when the system uses little endian, 0 otherwise.
```

Comment:
```suggestion
- endian flag: 1 when data is encoded by little endian, 0 otherwise.
```

Reply:
This flag is the system endian flag instead of the data. Most data are encoded in little endian. Some arrays such as `long[]` array may use native endian under the hood, we didn't change it to little endian. So if the peer has same endian, we can save the endian conversion cost

Reply:
Ahh I see. Thank you.

Reply:
> Ahh I see. Thank you.

Previously I think that actually there are other endian types rather than big and little, so we'd better describe explicitly.